src/renderers/server/ReactServerRenderer.js

```diff
@@ -0,0 +1,709 @@
+        if (defaultValue == null) {
+          defaultValue = '';
+        }
+        initialValue = defaultValue;
+      }
+
+      props = Object.assign({}, props, {
+        value: undefined,
+        children: '' + initialValue,
+      });
+    } else if (tag === 'select') {
+      if (__DEV__) {
+        ReactControlledValuePropTypes.checkPropTypes(
+          'select',
+          props,
+          () => '', // getCurrentFiberStackAddendum,
+        );
+
+        for (var i = 0; i < valuePropNames.length; i++) {
+          var propName = valuePropNames[i];
+          if (props[propName] == null) {
+            continue;
+          }
+          var isArray = Array.isArray(props[propName]);
+          if (props.multiple && !isArray) {
+            warning(
+              false,
+              'The `%s` prop supplied to <select> must be an array if ' +
+                '`multiple` is true.%s',
+              propName,
+              '', // getDeclarationErrorAddendum(),
+            );
+          } else if (!props.multiple && isArray) {
+            warning(
+              false,
+              'The `%s` prop supplied to <select> must be a scalar ' +
+                'value if `multiple` is false.%s',
+              propName,
+              '', // getDeclarationErrorAddendum(),
+            );
+          }
+        }
+      }
+
+      if (
+        props.value !== undefined &&
+        props.defaultValue !== undefined &&
+        !didWarnDefaultSelectValue
+      ) {
+        warning(
+          false,
+          'Select elements must be either controlled or uncontrolled ' +
+            '(specify either the value prop, or the defaultValue prop, but not ' +
+            'both). Decide between using a controlled or uncontrolled select ' +
+            'element and remove one of these props. More info: ' +
+            'https://fb.me/react-controlled-components',
+        );
+        didWarnDefaultSelectValue = true;
+      }
+
+      this.currentSelectValue = props.value != null
+        ? props.value
+        : props.defaultValue;
+      props = Object.assign({}, props, {
+        value: undefined,
+      });
+    } else if (tag === 'option') {
+      var selected = null;
+      var selectValue = this.currentSelectValue;
+      var optionChildren = flattenOptionChildren(props.children);
+      if (selectValue != null) {
+        var value;
+        if (props.value != null) {
+          value = props.value + '';
+        } else {
+          value = optionChildren;
+        }
+        selected = false;
+        if (Array.isArray(selectValue)) {
+          // multiple
+          for (var j = 0; j < selectValue.length; j++) {
+            if ('' + selectValue[j] === value) {
+              selected = true;
+              break;
+            }
+          }
+        } else {
+          selected = '' + selectValue === value;
+        }
+
+        props = Object.assign(
+          {
+            selected: undefined,
+            children: undefined,
```

Comment:
Is there any (perf) benefit to pass same shape objects as arguments of `Object.assign`?

Reply:
Likely depends on the engine and how it implements hidden class optimizations. In theory it helps provide a better layout for the fields.